client/client.go

```diff
@@ -1403,13 +1425,18 @@ func (c *Client) nodeID() (id, secret string, err error) {
 
 // setupNode is used to setup the initial node
 func (c *Client) setupNode() error {
-	node := c.config.Node
+	c.configLock.Lock()
+	defer c.configLock.Unlock()
+
+	newConfig := c.config.Copy()
```

Comment:
Are we holding the lock but also writing a copy of the config because we can't ensure nothing else is modifying the original config?

Reply:
Yes and more! The lock is doing two things in this case:

1. Go Memory Model Data Race Protection - Ensuring `c.config.Copy()` doesn't interleave with mutations in other threads (either to the c.config pointer following or the reading of the config struct itself).
2. Application Logic Race Protection - Ensuring that the mutations to `config` in this thread don't interleave with mutations in other threads (although in `setupNode()` this should never happen... although tests do strange things and if they can't be written race free, we can't use `go test -race` to detect races in non-test-code!).

We could satisfy (1) with fine grained locking like:

```go
func (c *Client) setupNode() error {
  c.configLock.Lock()
  newConfig := c.config.Copy()
  c.configLocik.Unlock()

  // ... Mutate newConfig safely here ...

  c.configLock.Lock()
  c.config = newConfig
  c.configLock.Unlock()
  return nil
}
```

Way smaller critical area and Go's race detector would still be happy! However (2) would be broken if there was another goroutine calling `UpdateConfig(func(c *config.Config) { c.Foo = bar })` during the `// ... Mutate ...` comment's work, its changes would be overwritten by the final setting of `c.config = newConfig`!

So we end up with big critical sections to ensure both Go's memory model and our application's logic are coherent and happy.

(Which brings us to your other comment... surely there's a better way?!)